lib/lib_rover/include/rover_helpers/PID.hpp

```diff
@@ -0,0 +1,108 @@
+#ifndef __PID_HPP__
+#define __PID_HPP__
+
+#include <Arduino.h>
+#include "rover_helpers/macros.hpp"
+#include "rover_helpers/log.hpp"
+
+class PID
+{
+public:
+    PID(float kp_, float ki_, float kd_, float intergalLimit_);
+    ~PID(void) {}
+    void init(void);
+    void setGains(float kp_, float ki_, float kd_);
+    void setIntLimit(float limit_);
+    float computeCommand(float error);
+    void reset(void);
+
+private:
+    float _kp = 0.0f;
+    float _ki = 0.0f;
+    float _kd = 0.0f;
+
+    float _cmdI = 0.0f;
+    float _integralLimit = 0.0f;
+    float _previousError = 0.0f;
+    float _lastMeasureTime = 0.0f;
+};
+
+PID::PID(float kp_, float ki_, float kd_, float intergalLimit_)
+{
+    this->setGains(kp_, ki_, kd_);
+    this->setIntLimit(intergalLimit_);
+    this->reset();
+    _lastMeasureTime = micros();
+}
+
+void PID::init(void)
+{
+    this->reset();
+}
+
+void PID::setGains(float kp_, float ki_, float kd_)
+{
+    _kp = kp_;
+    _ki = ki_;
+    _kd = kd_;
+}
+
+void PID::setIntLimit(float limit_)
+{
+    if (limit_ < 0.0f)
+    {
+        LOG(ERROR, "Integral limit should always be positive, abs value will be used");
+    }
+
+    _integralLimit = abs(limit_);
+}
+
+float PID::computeCommand(float error_)
+{
+    if (isnan(error_))
+    {
+        error_ = 0.0f;
+    }
+    if (isinf(error_))
+    {
+        error_ = 0.0f;
+    }
+
+    float currentTime = micros();
+    float dt = currentTime - _lastMeasureTime;
+
+    _cmdI += _ki*error_;
+
+    float cmdP = _kp * error_;
+    _cmdI = constrain(_cmdI, -_integralLimit, _integralLimit);
+    float cmdD = _kd * (error_ - _previousError) / (dt / 1'000'000.0f);
+
+    if (isnan(cmdP))
+    {
+        cmdP = 0.0f;
+    }
+    if (isnan(_cmdI))
+    {
+        _cmdI = 0.0f;
+    }
+    if (isnan(cmdD))
+    {
+        cmdD = 0.0f;
+    }
+
+    _previousError = error_;
+    _lastMeasureTime = currentTime;
+
+    // LOG(INFO, "cmdP: %.3f + cmdI: %.3f + cmdD: %.3f = %.3f", cmdP, _cmdI, cmdD, cmdP + _cmdI + cmdD);
+    // LOG(INFO, "Error: %.3f", error);
```

Comment:
Maybe rm comments

Reply:
I prefer we keep them because they'll be uncommented pretty much each time we're tuning PIDs.